File: 50. Pow(x, n)/50. Pow(x, n).rb
# @param {Float} x
# @param {Integer} n
# @return {Float}
def my_pow(x, n)
    return 1 if n == 0
    if n == -2**31
        x = 1 / x
        n = -(n + 1)
        return x * my_pow(x, n)
    end
    if n < 0
        x = 1 / x
        n = -n
    end
    result = 1.0
    while n > 0
        result *= x if n & 1 == 1
        x *= x
        n >>= 1
    end
    result
end

File: 50. Pow(x, n)/50. Pow(x, n).java
class Solution {
    public double myPow(double x, int n) {
        if (n == 0) return 1;
        if (n == Integer.MIN_VALUE) {
            x = 1 / x;
            n = -(n + 1);
            return x * myPow(x, n);
        }
        if (n < 0) {
            x = 1 / x;
            n = -n;
        }
        double result = 1;
        while (n > 0) {
            if ((n & 1) == 1) result *= x;
            x *= x;
            n >>= 1;
        }
        return result;
    }
}